### ISSUE TITLE
Pagination accessibility defects

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/vets-design-system-documentation#1408]
# Bug Report

- [x] I’ve searched for any related issues and avoided creating a duplicate issue.

## What happened

Several major accessibility issues exist with the pagination component (and documentation) including:

- Incompatibility with voice control software
- Lack of accessibility documentation for...
  - Focus management or use of `aria-live` 
  - Labelling paginations uniquely

### Compatibility with voice control software
Below is a demo of using the pagination using my macbook's voice control software. I'm using the following commands:
- **Click "x"** the easiest and most straightforward way in selecting something
- **"Show numbers"** the software shows numbers attached to all interactive controls to click on things that the first option can't achieve
- **"Show grid"** if show numbers doesn't work, show grid is often used as a last resort where a user can continue selecting squares within a grid until they reach their click target

https://user-images.githubusercontent.com/14154792/208672213-fe3abb4c-d333-4d41-b597-671db25f8a3d.mov

### Users cannot click pages directly due to `aria-labels` that append hidden content _before_ the visible string
If a user sees the number "3", they are likely to issue the command "Click 3"-- which will fail as we use an `aria-label` that appends an invisible "Page" in front of "3". 

Another issue is in how we abbreviate "Previous" to "Prev". For users with compounded disabilities (consider a case of someone with cognitive disabilities using voice recognition), it may be challenging to determine what an abbreviation stands for. 

For more, see [WCAG technique G208](https://www.w3.org/WAI/WCAG21/Techniques/general/G208)

> In most situations, where it is felt that additional context is needed, it is recommended that the visible text should precede the additional text. When authors make sure that the visible label of a control is included, intact, in the accessible name of that control, speech input users can be confident that their input will be correctly interpreted.

### Users cannot click "..." ellipses
[Eric Bailey documents issues related to ellipses](https://www.smashingmagazine.com/2022/06/voice-control-usability-considerations-partially-visually-hidden-link-names/#what-about-truncated-text-that-uses-an-ellipses), which may only be accessible through show numbers or the grid.

I replicated the same issue where no known voice command could select the ellipses.

### Missing accessibility documentation
A key issue that may be addressed through accessibility documentation or through further research is how we communicate changes in state through the current pagination system. As these are buttons (and not links), it is unclear how we inform screen reader users that they have successfully navigated to a new page.

For example, can we use an `aria-live` region somewhere on the page to indicate a change in how many results are currently displaying? Or should that be conducted through focus management (at the con of navigating the user away from the pagination on click)?

#### Assumption on links vs buttons
Ideally these should be links (see [govuk](https://design-system.service.gov.uk/components/pagination/), [W3C](https://design-system.w3.org/components/pagination.html), [nhs](https://service-manual.nhs.uk/design-system/components/pagination)) because to the user (at least perceptually) they are navigating page by page. 

But I believe (assumption) there was an engineering constraint that prevented us from doing this (but can’t confirm this without documentation on past efforts as I wasn’t on the team that worked on this).

### Minor bug on voiceover/safari/ios

https://user-images.githubusercontent.com/14154792/209204190-100a2146-69ed-42c5-b842-65d3e95b7f1f.MP4

Hover state/selected state appears for two pages at the same time.

## Alternatives
It may be worth considering simplified options for this component e.g. a "Next" and "Previous" option alone a la gov uk and NHS for easier controls.

<img width="159" alt="Screen Shot 2022-12-20 at 8 20 14 AM" src="https://user-images.githubusercontent.com/14154792/208676771-06da1d61-2ebc-4147-900d-6b27f5e8daf0.png">

This could be enhanced by including more plain language commands within the link/button text e.g.

"Previous page, 1 of 5"
"Next page, 3 of 5"

If we're worried about not giving users the option to be able to navigate to a specific page (a la a google search result), we may want to consider the following questions:
- Are there better, more direct solutions to addressing the need for navigating to a specific page? 
- If the goal is to consistently find a result, could that be better solved for through bookmarks, advanced search, filters, or sorts?

## Reproducing
- Formation version:
- Device: Macbook
- Browser: Safari

Steps to reproduce:

1. Turn on voice controls in your accessibility settings
2. Issue the following commands on the pagination
3. "Click 3" (confirm this fails)
4. "Click next" (confirm this fails) [this is a bit of an oddity, technically this should work, but I think since "next" is such a generic label, it may be conflating it with a natural voice control command
5. "Click prev" (confirm this fails)
6. "Click previous" (confirm this succeeds)
7. "Click dot dot dot" (confirm this fails)
8. "Click page 3" (confirm this fails) [this is also an oddity, as it is the full `aria-label`, again wonder if this is being conflated with a natural voice control command]
9. Edit the aria label from "page 3" to "3 page"
10. "Click 3 page" (confirm this succeeds) [swapping the order made this work, unsure why, needs additional testing]

## Urgency
How urgent is this request? Please select the appropriate option below and/or provide details
- [ ] This bug is blocking work currently in progress
- [ ] This bug is affecting work currently in progress but we have a workaround
- [x] This bug is blocking work planned within the next few sprints
- [ ] This bug is not blocking any work
- [ ] Other

**Comment by SarahKay8** [comment]
Hi @joshkimux can you specify which exact work items these defects prohibit?

**Comment by joshkimux** [comment]
@SarahKay8 yes, this would prohibit https://github.com/department-of-veterans-affairs/va.gov-team/issues/17149

**Comment by SarahKay8** [comment]
Thanks! cc: @jamigibbs @caw310

**Comment by jamigibbs** [comment]
@joshkimux Thank you for all of your feedback here! As I'm starting to work through it and I'm hoping to clarifying the markup.

### button vs link

> Ideally these should be links (see [govuk](https://design-system.service.gov.uk/components/pagination/), [W3C](https://design-system.w3.org/components/pagination.html), [nhs](https://service-manual.nhs.uk/design-system/components/pagination)) because to the user (at least perceptually) they are navigating page by page.

> But I believe (assumption) there was an engineering constraint that prevented us from doing this (but can’t confirm this without documentation on past efforts as I wasn’t on the team that worked on this).

In the [original React component](https://design.va.gov/storybook/?path=/docs/deprecated-pagination-react--default) (deprecated), the markup was using anchor links `a`.

Looking at the history of this change, it appears that it was updated to `button` when the component was converted to a web component with the [intention of improving screenreader support](https://github.com/department-of-veterans-affairs/component-library/pull/275#discussion_r802194853). It does not appear there were any actual technical constraints (except to relieve us of a little extra work handling empty `href` values)

From the previous developer:

> Using NVDA, the React component reads out as "Link previous page, link page 1, link page 2, ...link next page". This web component with the markup changes reads out as "Navigation landmark list with 1 item: button previous page, out of list... list with 5 items: button page 1, button page 2, ...out of list, list with 1 items: button next page". I think this adds a lot of value to screen readers.

So I just wanted to double check to make sure I wasn't introducing any other issues with screenreaders if we went back to links instead of buttons based on the comments from the previous developer.

**Comment by joshkimux** [comment]
@jamigibbs got it! This is really great context. 

If I'm reading this right: because these are react components, the `a` anchor links have empty `href` values meaning that they won't take on a natural tab stop (hence switching to buttons instead of setting all of their tabindex's to 0).

If we were to go back to `a` anchor links, we'd ideally be able to:
1. populate the empty `href`s with actual URLs (so we don't have to rely on tabindex) and
2. style the numbers as links (as opposed to buttons)

I'm not an engineer by any means-- but I'm guessing there may be a constraint related to react where we can't populate the `hrefs` with actual URLs. Is that correct?

As for the bit on adding value to screen readers, I think the developer was commenting on placing it within a `nav` landmark and an unordered/ordered list element which makes sense ✔️

**Comment by jamigibbs** [comment]
@joshkimux I don't think the `href` limitation is React specific. If we went back to links in the web component, we would need to add `tabIndex=0` to that version too. The component is designed to just return a page number when a page is clicked (1, 2, 3, etc.) instead of triggering an actual page navigation (which is what href would do). The developer implementing the component would handle the action based on the page number (rather than the component handling the action).

At some point I think we need to change this approach in our component though as we move towards converting them to their [USWDS counterpart](https://designsystem.digital.gov/components/pagination/) (which also uses links just like all of your other examples). 

But I agree that the links should be styled as links if that's the markup we are going to use. I'll make sure that is added to my list and connect with our designer to give a green light on what we end up with.

For now though (and to get you unblocked in your work with the voice control concerns), I wonder if we could either continue using `button` (to avoid `tabIndex=0`) or switch back to links (and use `tabIndex=0`)?

**Comment by joshkimux** [comment]
That sounds like a good plan! I think sticking with buttons (to avoid `tabIndex=0`) and to avoid having to do extra work in the meantime if a more thorough fix will occur in the near future.

That leaves the main issues documented in the ticket around `aria-labels`-- particularly for the "..." ellipses which is a bit of a tough problem to get around. Does the DST have an a11y resource that can support there?

**Comment by caw310** [comment]
@joshkimux , actually the DST no longer has an a11y specialist. Sarah was moved on to another contract. If you could help out with this issue, that would be much appreciated.

**Comment by joshkimux** [comment]
@caw310 I really want to! I don't have the bandwidth at the moment, and am currently filling in some gaps as we've lost a couple of a11y specialists. Hopefully once the transition of a11y specialists smoothes out we can find someone 🙇

**Comment by caw310** [comment]
Carrying over into sprint 19 to fix some of the problems identified. 
We do need a11y support to evaluate if USWDS meets our needs or not.

**Comment by jamigibbs** [comment]
### Voice Commands

While I was able to get Voice Command working for the following commands, I do want to mention that the software seems to be very sensitive to the enunciation of the words. Sometimes I needed to say the command multiple times with very clear enunciation before it worked.

Voice Commands that are working:

- "Click page 3" ✅  (and other page specific numbers)
- "Click previous page" ✅ 
- "Click next page" ✅ 

The command structure that I was unable to make work:

- "Click 3" ❌ 

I think that this type of command might not have enough context for Voice Command to understand where to click. I was also not able to get this command to work in the [govuk](https://design-system.service.gov.uk/), [uswds](https://designsystem.digital.gov/components/pagination/), or [w3](https://design-system.w3.org/components/pagination.html) design system pagination components.

The use of "page" as part of the voice command seems to be consistent with the other pagination components though and having this value as part of `aria-label` seemed to help trigger it.

### Ellipses (...) - Click behavior

In our VA component, we were previously allowing for click behavior to happen within the code (even though it didn't actually do anything). After looking at how other systems were accounting for this type of treatment, I noticed that there were a few things we could do to make it less confusing:

1. Add the `role="presentation"` to the element.
2. Make sure that when tabbing through the list that the elipses were skipped.

This seems to follow what these examples are doing as well:

### [govuk](https://design-system.service.gov.uk/)
![Screenshot 2023-01-13 at 3 01 20 PM](https://user-images.githubusercontent.com/872479/212418321-d596594b-58f7-4141-be5d-cb7e0a2bf3d0.png)

### [uswds](https://designsystem.digital.gov/components/pagination/)
![Screenshot 2023-01-13 at 3 01 11 PM](https://user-images.githubusercontent.com/872479/212418323-edfa131a-6b46-4df8-8b93-1102e892c232.png)

### [w3](https://design-system.w3.org/components/pagination.html)
![Screenshot 2023-01-13 at 3 01 04 PM](https://user-images.githubusercontent.com/872479/212418325-6f410409-7f85-46f1-8d6f-1c5f0eabb6bc.png)

### Ellipses (...) - Display

If there's concern about displaying the ellipses in general (because we feel like it might be obfuscating information), it is possible right now to remove the ellipses by setting the `show-last-page` to `false`. In this case, `va-pagination` will not show the last page and will simply tab one page at a time with the "next" button.

`show-last-page=true`
![Screenshot 2023-01-13 at 3 13 56 PM](https://user-images.githubusercontent.com/872479/212420537-449e1283-711a-404a-9aa1-b3a69cddc7df.png)

`show-last-page=false`
![Screenshot 2023-01-13 at 3 13 51 PM](https://user-images.githubusercontent.com/872479/212420594-d7800442-185f-4190-9c4a-138bfbfc50cb.png)

I connected with @GnatalieH to discuss the ellipses treatment and after reviewing the other examples above (govuk, uswds, w3), our conclusion was that the ellipses aren't truncating any text, they seem to be just presentational, and are not meant to be clicked or used functionally so we will likely not be redesigning this approach.

### Button style vs Link Style

At the moment, our pagination component is using `button` for reasons discussed above. So they are styled as buttons. When we start using the [USWDS pagination component](https://designsystem.digital.gov/components/pagination/) though which uses links, we would expect those to be styled as links. This might be some feedback to provide the USWDS team ahead of our consumption of their modules that the current styles should  probably be updated to look more link links rather than buttons (ie. add an underline to the text, remove the border on the inactive links).

### [uswds](https://designsystem.digital.gov/components/pagination/)
![Screenshot 2023-01-13 at 3 01 11 PM](https://user-images.githubusercontent.com/872479/212418323-edfa131a-6b46-4df8-8b93-1102e892c232.png)

### Next Steps

- [ ] Confirm that Voice Command is functioning as part of [this PR ](https://github.com/department-of-veterans-affairs/component-library/pull/602) - @joshkimux would you be able to validate this from the [PR's Chromatic link](https://1408-pagination-a11y--60f9b557105290003b387cd5.chromatic.com/?path=/docs/components-va-pagination--default)?
- [ ] Add or discuss the addition of a11y documentation to [design.va.gov for `va-pagination`](https://design.va.gov/components/pagination). Josh has mentioned adding clarity around focus management or use of aria-live as well as labelling paginations uniquely. - @caw310 I'm not sure if this would require a separate ticket for our team or how we would like to manage this request considering a pending USWDS transition.
- [ ] When an a11y resource is available to the design system, we should do our own a11y audit on the [USWDS pagination component](https://designsystem.digital.gov/components/pagination/) ahead of us transitioning to it. @caw310 I'm not sure when this can get done considering DST resourcing so just pinging you for awareness.

**Comment by joshkimux** [comment]
@jamigibbs this is an amazing writeup, I appreciate the time you took to do this ⭐ 

I squeezed out [some time to do some testing which I've verified here](https://github.com/department-of-veterans-affairs/component-library/pull/602#issuecomment-1386071376) 🙏 

Just a note here that we may need a specialist with access to Dragon Naturally Speaking to test this as well (as it's the more commonly used voice control software). Is there any way to include that in the accessibility documentation?

Also wanted to follow up from a design perspective (@GnatalieH) on if there's any possibility we can explore plain language pagination alternatives listed above that don't involve a large clickable list of numbers (not in all cases, but likely in a number of contexts to reduce the amount of interactive elements and make voice interaction a little easier).

**Comment by GnatalieH** [comment]
Hi @joshkimux. I like the plain language examples you shared, but aren't they only appropriate for five or fewer pages? Not providing a way to skip from, say, page 1 to page 10 (without clicking through each page linearly) isn't a great user experience. What did you have in mind? Thanks!

**Comment by joshkimux** [comment]
@GnatalieH Great points, some thoughts on that:
- Ideally, I think from a content perspective, we really shouldn't be having results of X+ number of pages to go through for most pages. I feel like it's more of a symptom of filtering, search, and IA in need of improvement. Having to guess what will exist on what page, instead of being able to effectively filter/sort/search what you need to the top of the list (or not have to do so at all if the content is reduced to what the user needs) feels like an area we can improve on.
- Given that, I wonder if pagination should default to a more plain language style like this, with the fall back experience being the use of our current pagination system which is designed for ~5+ pages.

It's definitely a future state design, but I'd love to see teams that have less than 4 or so pages explore it as an option 😄

**Comment by GnatalieH** [comment]
@joshkimux I agree with every point you made. I can create a ticket to explore plain language pagination options for a smaller number of pages. And hopefully we can improve our search/filtering options, too. I agree that we shouldn't have the scenario where the user is exposed to pages and pages of content to parse through.

**Comment by GnatalieH** [comment]
@joshkimux I created [this ticket](https://app.zenhub.com/workspaces/platform-design-system-5f8de67192551b0012ebb802/issues/gh/department-of-veterans-affairs/vets-design-system-documentation/1463) as a follow up to explore a plain language variation of the Pagination component. Please feel free to make edits to or comments on the ticket.

I'm going to close this ticket, point to the [follow up ticket](https://app.zenhub.com/workspaces/platform-design-system-5f8de67192551b0012ebb802/issues/gh/department-of-veterans-affairs/vets-design-system-documentation/1463), and suggest that we be on the lookout for ways to improve our content/IA and search/filter/sort functionality to help mitigate instances of pages and pages of content that the user has to click through in order to find what they're looking for.

**Comment by joshkimux** [comment]
This is amazing. Thank you so much @GnatalieH and @jamigibbs , y'all are the dream team ❤️ ❇️
